File: face_detc.cpp
void generatePriors()
{
    // Calculate shapes of different scales according to the shape of input image
    Size feature_map_2nd = {
        int(int((inputW+1)/2)/2), int(int((inputH+1)/2)/2)
    };
    Size feature_map_3rd = {
        int(feature_map_2nd.width/2), int(feature_map_2nd.height/2)
    };
    Size feature_map_4th = {
        int(feature_map_3rd.width/2), int(feature_map_3rd.height/2)
    };
    Size feature_map_5th = {
        int(feature_map_4th.width/2), int(feature_map_4th.height/2)
    };
    Size feature_map_6th = {
        int(feature_map_5th.width/2), int(feature_map_5th.height/2)
    };

    std::vector<Size> feature_map_sizes;
    feature_map_sizes.push_back(feature_map_3rd);
    feature_map_sizes.push_back(feature_map_4th);
    feature_map_sizes.push_back(feature_map_5th);
    feature_map_sizes.push_back(feature_map_6th);

    // Fixed params for generating priors
    const std::vector<std::vector<float>> min_sizes = {
        {10.0f,  16.0f,  24.0f},
        {32.0f,  48.0f},
        {64.0f,  96.0f},
        {128.0f, 192.0f, 256.0f}
    };
    CV_Assert(min_sizes.size() == feature_map_sizes.size()); // just to keep vectors in sync
    const std::vector<int> steps = { 8, 16, 32, 64 };

    // Generate priors
    priors.clear();
    for (size_t i = 0; i < feature_map_sizes.size(); ++i)
    {
        Size feature_map_size = feature_map_sizes[i];
        std::vector<float> min_size = min_sizes[i];

        for (int _h = 0; _h < feature_map_size.height; ++_h)
        {
            for (int _w = 0; _w < feature_map_size.width; ++_w)
            {
                for (size_t j = 0; j < min_size.size(); ++j)
                {
                    float s_kx = min_size[j] / inputW;
                    float s_ky = min_size[j] / inputH;

                    float cx = (_w + 0.5f) * steps[i] / inputW;
                    float cy = (_h + 0.5f) * steps[i] / inputH;

                    Rect2f prior = { cx, cy, s_kx, s_ky };
                    priors.push_back(prior);
                }
            }
        }
    }
}

Mat postProcess(const std::vector<Mat>& output_blobs)
{
    // Extract from output_blobs
    Mat loc = output_blobs[0];
    Mat conf = output_blobs[1];
    Mat iou = output_blobs[2];

    // Decode from deltas and priors
    const std::vector<float> variance = {0.1f, 0.2f};
    float* loc_v = (float*)(loc.data);
    float* conf_v = (float*)(conf.data);
    float* iou_v = (float*)(iou.data);
    Mat faces;
    // (tl_x, tl_y, w, h, re_x, re_y, le_x, le_y, nt_x, nt_y, rcm_x, rcm_y, lcm_x, lcm_y, score)
    // 'tl': top left point of the bounding box
    // 're': right eye, 'le': left eye
    // 'nt':  nose tip
    // 'rcm': right corner of mouth, 'lcm': left corner of mouth
    Mat face(1, 15, CV_32FC1);
    for (size_t i = 0; i < priors.size(); ++i) {
        // Get score
        float clsScore = conf_v[i*2+1];
        float iouScore = iou_v[i];
        // Clamp
        if (iouScore < 0.f) {
            iouScore = 0.f;
        }
        else if (iouScore > 1.f) {
            iouScore = 1.f;
        }
        float score = std::sqrt(clsScore * iouScore);
        face.at<float>(0, 14) = score;

        // Get bounding box
        float cx = (priors[i].x + loc_v[i*14+0] * variance[0] * priors[i].width)  * inputW;
        float cy = (priors[i].y + loc_v[i*14+1] * variance[0] * priors[i].height) * inputH;
        float w  = priors[i].width  * exp(loc_v[i*14+2] * variance[0]) * inputW;
        float h  = priors[i].height * exp(loc_v[i*14+3] * variance[1]) * inputH;
        float x1 = cx - w / 2;
        float y1 = cy - h / 2;
        face.at<float>(0, 0) = x1;
        face.at<float>(0, 1) = y1;
        face.at<float>(0, 2) = w;
        face.at<float>(0, 3) = h;

        // Get landmarks
        face.at<float>(0, 4) = (priors[i].x + loc_v[i*14+ 4] * variance[0] * priors[i].width)  * inputW;  // right eye, x
        face.at<float>(0, 5) = (priors[i].y + loc_v[i*14+ 5] * variance[0] * priors[i].height) * inputH;  // right eye, y
        face.at<float>(0, 6) = (priors[i].x + loc_v[i*14+ 6] * variance[0] * priors[i].width)  * inputW;  // left eye, x
        face.at<float>(0, 7) = (priors[i].y + loc_v[i*14+ 7] * variance[0] * priors[i].height) * inputH;  // left eye, y
        face.at<float>(0, 8) = (priors[i].x + loc_v[i*14+ 8] * variance[0] * priors[i].width)  * inputW;  // nose tip, x
        face.at<float>(0, 9) = (priors[i].y + loc_v[i*14+ 9] * variance[0] * priors[i].height) * inputH;  // nose tip, y
        face.at<float>(0, 10) = (priors[i].x + loc_v[i*14+10] * variance[0] * priors[i].width)  * inputW; // right corner of mouth, x
        face.at<float>(0, 11) = (priors[i].y + loc_v[i*14+11] * variance[0] * priors[i].height) * inputH; // right corner of mouth, y
        face.at<float>(0, 12) = (priors[i].x + loc_v[i*14+12] * variance[0] * priors[i].width)  * inputW; // left corner of mouth, x
        face.at<float>(0, 13) = (priors[i].y + loc_v[i*14+13] * variance[0] * priors[i].height) * inputH; // left corner of mouth, y

        faces.push_back(face);
    }

    if (faces.rows > 1)
    {
        // Retrieve boxes and scores
        std::vector<Rect2i> faceBoxes;
        std::vector<float> faceScores;
        for (int rIdx = 0; rIdx < faces.rows; rIdx++)
        {
            faceBoxes.push_back(Rect2i(int(faces.at<float>(rIdx, 0)),
                                        int(faces.at<float>(rIdx, 1)),
                                        int(faces.at<float>(rIdx, 2)),
                                        int(faces.at<float>(rIdx, 3))));
            faceScores.push_back(faces.at<float>(rIdx, 14));
        }

        std::vector<int> keepIdx;
        dnn::NMSBoxes(faceBoxes, faceScores, scoreThreshold, nmsThreshold, keepIdx, 1.f, topK);

        // Get NMS results
        Mat nms_faces;
        for (int idx: keepIdx)
        {
            nms_faces.push_back(faces.row(idx));
        }
        return nms_faces;
    }
    else
    {
        return faces;
    }
}

File: face_detc.py
import numpy as np
import cv2 as cv

def generate_priors( input_w, input_h):
    # Calculate shapes of different scales according to the shape of input image
    feature_map_2nd = (int((input_w + 1) / 2) // 2, int((input_h + 1) / 2) // 2)
    feature_map_3rd = (int(feature_map_2nd[0] / 2), int(feature_map_2nd[1] / 2))
    feature_map_4th = (int(feature_map_3rd[0] / 2), int(feature_map_3rd[1] / 2))
    feature_map_5th = (int(feature_map_4th[0] / 2), int(feature_map_4th[1] / 2))
    feature_map_6th = (int(feature_map_5th[0] / 2), int(feature_map_5th[1] / 2))

    feature_map_sizes = [feature_map_3rd, feature_map_4th, feature_map_5th, feature_map_6th]

    # Fixed params for generating priors
    min_sizes = [[10.0,  16.0,  24.0],
                 [32.0,  48.0],
                 [64.0,  96.0],
                 [128.0, 192.0, 256.0]]
    steps = [8, 16, 32, 64]

    # Generate priors
    priors = []
    for i in range(len(feature_map_sizes)):
        feature_map_size = feature_map_sizes[i]
        min_size = min_sizes[i]

        for _h in range(feature_map_size[1]):
            for _w in range(feature_map_size[0]):
                for j in range(len(min_size)):
                    s_kx = min_size[j] / input_w
                    s_ky = min_size[j] / input_h

                    cx = (_w + 0.5) * steps[i] / input_w
                    cy = (_h + 0.5) * steps[i] / input_h

                    prior = (cx, cy, s_kx, s_ky)
                    priors.append(prior)
    return priors


def post_process(output_blobs, input_w, input_h):

    # Extract from output_blobs
    loc = output_blobs[0]
    conf = output_blobs[1]
    iou = output_blobs[2]

    # Decode from deltas and priors
    variance = [0.1, 0.2]
    loc_v = loc.flatten()
    conf_v = conf.flatten()
    iou_v = iou.flatten()
    faces = []
    # (tl_x, tl_y, w, h, re_x, re_y, le_x, le_y, nt_x, nt_y, rcm_x, rcm_y, lcm_x, lcm_y, score)
    # 'tl': top left point of the bounding box
    # 're': right eye, 'le': left eye
    # 'nt':  nose tip
    # 'rcm': right corner of mouth, 'lcm': left corner of mouth
    face = np.empty((1, 15))
    priors = generate_priors( input_w, input_h)
    for i in range(len(priors)):
        # Get score
        cls_score = conf_v[i*2+1]
        iou_score = iou_v[i]
        # Clamp
        if iou_score < 0:
            iou_score = 0
        elif iou_score > 1:
            iou_score = 1
        score = np.sqrt(cls_score * iou_score)
        face[0, 14] = score

        # Get bounding box
        cx = (priors[i][0] + loc_v[i*14+0] * variance[0] * priors[i][2]) * input_w
        cy = (priors[i][1] + loc_v[i*14+1] * variance[0] * priors[i][3]) * input_h
        w  = priors[i][2] * np.exp(loc_v[i*14+2] * variance[0]) * input_w
        h  = priors[i][3] * np.exp(loc_v[i*14+3] * variance[1]) * input_h
        x1 = cx - w / 2
        y1 = cy - h / 2
        face[0, 0] = x1
        face[0, 1] = y1
        face[0, 2] = w
        face[0, 3] = h

        # Get landmarks
        face[0, 4] = (priors[i][0] + loc_v[i*14+4] * variance[0] * priors[i][2]) * input_w  # right eye, x
        face[0, 5] = (priors[i][1] + loc_v[i*14+5] * variance[0] * priors[i][3]) * input_h  # right eye, y
        face[0, 6] = (priors[i][0] + loc_v[i*14+6] * variance[0] * priors[i][2]) * input_w  # left eye, x
        face[0, 7] = (priors[i][1] + loc_v[i*14+7] * variance[0] * priors[i][3]) * input_h  # left eye, y
        face[0, 8] = (priors[i][0] + loc_v[i*14+8] * variance[0] * priors[i][2]) * input_w  # nose tip, x
        face[0, 9] = (priors[i][1] + loc_v[i*14+9] * variance[0] * priors[i][3]) * input_h  # nose tip, y
        face[0, 10] = (priors[i][0] + loc_v[i*14+10] * variance[0] * priors[i][2]) * input_w  # right corner of mouth, x
        face[0, 11] = (priors[i][1] + loc_v[i*14+11] * variance[0] * priors[i][3]) * input_h  # right corner of mouth, y
        face[0, 12] = (priors[i][0] + loc_v[i*14+12] * variance[0] * priors[i][2]) * input_w  # left corner of mouth, x
        face[0, 13] = (priors[i][1] + loc_v[i*14+13] * variance[0] * priors[i][3]) * input_h  # left corner of mouth, y

        faces.append(face)

    if len(faces) > 1:
        # Retrieve boxes and scores
        face_boxes = []
        face_scores = []
        for r_idx in range(len(faces)):
            face_boxes.append((int(faces[r_idx][0, 0]),
                               int(faces[r_idx][0, 1]),
                               int(faces[r_idx][0, 2]),
                               int(faces[r_idx][0, 3])))
            face_scores.append(faces[r_idx][0, 14])

        keep_idx = []
        # dnn.NMSBoxes(face_boxes, face_scores, score_threshold, nms_threshold, keep_idx, 1.0, top_k)
        keep_idx = cv.dnn.NMSBoxes(face_boxes, face_scores, 0.5, 0.3, top_k=5000)
        # Get NMS results
        nms_faces = []
        print(len(keep_idx))
        print(priors[0])
        
        return 1

    #     for idx in keep_idx:
    #         nms_faces.append(faces[idx])
    #     return np.vstack(np.array(nms_faces))
    # else:
    #     return np.vstack(np.array(faces))

if __name__ == "__main__":

    conf = np.loadtxt("./data/conf.out")
    loc = np.loadtxt("./data/loc.out")
    iou = np.loadtxt("./data/iou.out")

    output_blobs = [loc, conf, iou]

    result = post_process(output_blobs, 160, 120)